ethcore/src/state/account.rs

```diff
@@ -457,16 +553,21 @@ impl Account {
 	pub fn overwrite_with(&mut self, other: Account) {
 		self.balance = other.balance;
 		self.nonce = other.nonce;
-		self.storage_root = other.storage_root;
 		self.code_hash = other.code_hash;
 		self.code_filth = other.code_filth;
 		self.code_cache = other.code_cache;
 		self.code_size = other.code_size;
 		self.address_hash = other.address_hash;
-		let mut cache = self.storage_cache.borrow_mut();
-		for (k, v) in other.storage_cache.into_inner() {
-			cache.insert(k, v);
+		if self.storage_root == other.storage_root {
+			let mut cache = self.storage_cache.borrow_mut();
+			for (k, v) in other.storage_cache.into_inner() {
+				cache.insert(k, v);
+			}
+		} else {
+			self.storage_cache = other.storage_cache;
```

Comment:
This makes sense to me (the replacement logic for new contracts and cache), but I do not have the background to be confident with this new behavior in the context of the evm. (I am not requesting anything, jut pointing out that my review is a bit incomplete on this part of the code).

Reply:
In this case, `self.storage_cache` does not equal `other.storage_cache`. This means one of them is `KECCAK_NULL_RLP`. We cannot merge the storage cache here. Because this function is to "overwrite" `self` with `other`, we just move `other`'s storage cache to `self`.

Reply:
Yes the logic is sound, thanks for confirming it, I was just worried about some 'assertion' that could have been made on existing code.